pages/submissionFile/submissionFile.js

```diff
@@ -0,0 +1,76 @@
+// @ts-check
+const express = require('express');
+const asyncHandler = require('express-async-handler');
+const { isBinaryFile } = require('isbinaryfile');
+const mime = require('mime');
+const sqldb = require('@prairielearn/postgres');
+
+const sql = sqldb.loadSqlEquiv(__filename);
+const router = express.Router({ mergeParams: true });
+
+const MEDIA_PREFIXES = ['image/', 'audio/', 'video/'];
+
+/**
+ * Guesses the mime type for a file based on its name and contents.
+ *
+ * @param {string} name The file's name
+ * @param {Buffer} buffer The file's contents
+ * @returns {Promise<string>} The guessed mime type
+ */
+async function guessMimeType(name, buffer) {
+  const mimeType = mime.getType(name);
+  const isBinary = await isBinaryFile(buffer);
+
+  if (!isBinary) {
+    return 'text/plain';
+  }
+
+  if (mimeType && MEDIA_PREFIXES.some((p) => mimeType.startsWith(p))) {
+    return mimeType;
+  }
+
+  return 'application/octet-stream';
+}
+
+router.get(
+  '/*',
+  asyncHandler(async (req, res) => {
+    const submissionId = req.params.submission_id;
+    const fileName = req.params[0];
+
+    const fileRes = await sqldb.queryZeroOrOneRowAsync(sql.select_submission_file, {
+      question_id: res.locals.question.id,
+      instance_question_id: res.locals.instance_question?.id ?? null,
+      submission_id: submissionId,
+      file_name: fileName,
+    });
+
+    if (fileRes.rowCount === 0) {
+      res.sendStatus(404);
+      return;
+    }
+
+    const contents = fileRes.rows[0].contents;
+    if (!contents) {
+      res.sendStatus(404);
+      return;
+    }
```

Comment:
Wouldn't this cause an empty file to cause an error? I know `pl-file-upload` does not allow for blank files, but other elements (or files submitted via workspaces) may be empty.

Reply:
Good call - I believe it would. Checking explicitly for `null` is probably better.

Reply:
I just created two empty files in a worksheet, here's the result:

```
postgres=# select submitted_answer from submissions where variant_id = 286;
                                                                                                                                                                  submitted_answer

----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 {"_files": [{"name": "empty.c", "contents": ""}, {"name": "empty.h", "contents": ""}, {"name": "starter_code.c", "contents": "Ly8gc3RhcnRlcl9jb2RlLmMKCiNpbmNsdWRlIDxzdGRpby5oPgojaW5jbHVkZSAic3RhcnRlcl9jb2RlLmgiCgppbnQgbWFpbigpCnsKICAgIC8vIGVudGVyIHlvdXIgY29kZQp9Cg=="}, {"name": "starter_code.h", "contents": "Ly8gc3RhcnRlcl9jb2RlLmgK"}]}
(1 row)
```

Reply:
Changed to check for `contents == null` in ca1ebda.